test/unit/tools/test_history_imp_exp.py

```diff
@@ -101,6 +109,131 @@ def test_history_import_abspath_in_metadata():
+    assert imported_job.output_datasets[0].dataset == datasets[1]
+
+    assert imported_job.input_datasets
+    assert imported_job.input_datasets[0].dataset == datasets[0]
+
+    assert datasets[0].state == 'ok'
+    assert datasets[1].state == 'ok'
+
+    with open(datasets[0].file_name, "r") as f:
+        assert f.read().startswith("chr1    4225    19670")
+    with open(datasets[1].file_name, "r") as f:
+        assert f.read().startswith("chr1\t147962192\t147962580\tNM_005997_cds_0_0_chr1_147962193_r\t0\t-")
+
+
+def test_export_dataset_with_deleted_and_purged():
+    app, sa_session, h = _setup_history_for_export("Datasets History with deleted")
+
+    d1, d2 = _create_datasets(sa_session, h, 2)
+
+    # Maybe use abstractions for deleting?
+    d1.deleted = True
+    d1.dataset.deleted = True
+    d1.dataset.purged = False
+
+    d2.deleted = True
+    d2.dataset.deleted = True
+    d2.dataset.purged = True
+
+    j1 = model.Job()
+    j1.user = h.user
+    j1.tool_id = "cat1"
+    j1.add_output_dataset("out_file1", d1)
+
+    j2 = model.Job()
+    j2.user = h.user
+    j2.tool_id = "cat1"
+    j2.add_output_dataset("out_file1", d2)
+
+    sa_session.add(d1)
+    sa_session.add(d2)
+    sa_session.add(j1)
+    sa_session.add(j2)
+    sa_session.add(h)
+    sa_session.flush()
+
+    assert d1.deleted
+
+    app.object_store.update_from_file(d1, file_name="test-data/1.txt", create=True)
+    app.object_store.update_from_file(d2, file_name="test-data/2.bed", create=True)
+
+    imported_history = _import_export(app, h)
+
+    datasets = list(imported_history.contents_iter(types=["dataset"]))
+    assert len(datasets) == 1
+
+    assert datasets[0].state == 'discarded'
+    assert datasets[0].deleted
+    assert datasets[0].dataset.deleted
+    assert datasets[0].creating_job
+
+
+def _create_datasets(sa_session, history, n):
+    return [model.HistoryDatasetAssociation(extension="txt", history=history, create_dataset=True, sa_session=sa_session, hid=i + 1) for i in range(n)]
+
+
+def _setup_history_for_export(history_name):
+    app = _mock_app()
+    sa_session = app.model.context
+
+    email = history_name.replace(" ", "-") + "-user@example.org"
+    u = model.User(email=email, password="password")
+    h = model.History(name=history_name, user=u)
+
+    return app, sa_session, h
+
+
+def _import_export(app, h, dest_export=None):
+    if dest_export is None:
+        dest_parent = mkdtemp()
+        dest_export = os.path.join(dest_parent, "moo.tgz")
+
+    jeha = model.JobExportHistoryArchive(job=None, history=h,
+                                         dataset=None,
+                                         compressed=True)
+    wrapper = JobExportHistoryArchiveWrapper(app, 1)
+    wrapper.setup_job(jeha)
+
+    from galaxy.tools.imp_exp import export_history
+    ret = export_history.main(["--gzip", jeha.temp_directory, dest_export])
+    assert ret == 0
+
+    _, imported_history = import_archive(dest_export, app=app)
+    print(_)
```

Comment:
Do you want to keep this in ?

Reply:
Nope, but I'll rebase a fix into #7367 - this PR is so green and approved 😆.

Reply:
Sounds good.